x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/analysis/MetricsTranslator.java

```diff
@@ -0,0 +1,151 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.analysis;
+
+import org.elasticsearch.index.IndexMode;
+import org.elasticsearch.xpack.esql.core.expression.Alias;
+import org.elasticsearch.xpack.esql.core.expression.Attribute;
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.expression.Expressions;
+import org.elasticsearch.xpack.esql.core.expression.MetadataAttribute;
+import org.elasticsearch.xpack.esql.core.expression.NamedExpression;
+import org.elasticsearch.xpack.esql.core.expression.UnresolvedAttribute;
+import org.elasticsearch.xpack.esql.core.index.EsIndex;
+import org.elasticsearch.xpack.esql.core.plan.logical.LogicalPlan;
+import org.elasticsearch.xpack.esql.expression.function.aggregate.AggregateFunction;
+import org.elasticsearch.xpack.esql.expression.function.aggregate.Rate;
+import org.elasticsearch.xpack.esql.expression.function.aggregate.Values;
+import org.elasticsearch.xpack.esql.expression.function.grouping.Bucket;
+import org.elasticsearch.xpack.esql.plan.logical.EsRelation;
+import org.elasticsearch.xpack.esql.plan.logical.EsqlAggregate;
+import org.elasticsearch.xpack.esql.plan.logical.metrics.UnresolvedMetrics;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Stream;
+
+/**
+ * Rate aggregation is special. Regardless of the grouping keys, it must be computed per time series.
+ * This means the keys must be `_tisd` or a pair of `_tisd` and `time_bucket`. To support user-defined grouping keys,
+ * we first execute the rate aggregation using the time-series keys, then perform another aggregation
+ * with the resulting rate using the user-specific keys.
+ * <p>
+ * This class translates the aggregates in the METRICS commands to standard STATS commands. This approach helps
+ * avoid introducing new plans and operators for metrics aggregations to support the METRICS command.
+ * However, we may need to gradually introduce these features for performance reasons.
+ */
+final class MetricsTranslator {
```

Comment:
This is the main change. Other changes have been spun off to https://github.com/elastic/elasticsearch/pull/109982 and https://github.com/elastic/elasticsearch/pull/109983.